packages/jitm/src/class-message.php

```diff
@@ -0,0 +1,1168 @@
+		$score = array_reduce(
+			array_keys( $query ),
+			function ( $score, $key ) use ( &$query ) {
+				if ( $score ) {
+					return $score;
+				}
+
+				if ( array_key_exists( $key, $this->query ) ) {
+					return (int) preg_match( $this->query[ $key ], $query[ $key ] );
+				}
+
+				return 0;
+			},
+			0
+		);
+
+		return $score ? $score : false;
+	}
+
+	/**
+	 * Score option matches
+	 *
+	 * @param string $path path.
+	 * @param array  $query query.
+	 * @param int    $score score.
+	 *
+	 * @return bool|int
+	 */
+	private function score_option_matches( $path, $query, $score ) {
+		if ( empty( $this->option_matches ) ) {
+			return $score;
+		}
+
+		// If any of the matches are rejected, reject the rule entirely.
+		if ( array_filter(
+			$this->option_matches,
+			function ( $match ) {
+				return false === $match;
+			}
+		) ) {
+			return false;
+		}
+
+		// If any of the matches are accepted with an integer, use the largest as the score.
+		$ints = array_filter( $this->option_matches, 'is_int' );
+		if ( $ints ) {
+			return max( $ints );
+		}
+
+		// Otherwise, accept the rule with score 1.
+		return 1;
+	}
+
+	/**
+	 * Score dismissal
+	 *
+	 * @param string $path path.
+	 * @param array  $query query.
+	 * @param int    $score score.
+	 *
+	 * @return bool|int
+	 */
+	private function score_dismissal( $path, $query, $score ) {
+		$dismissals = $this->get_dismissals();
+		if ( false !== $dismissals && is_array( $dismissals ) && isset( $dismissals[ $this->feature_class ] ) && is_array( $dismissals[ $this->feature_class ] ) ) {
+			$score = 0;
+
+			$dismissal = $dismissals[ $this->feature_class ];
+			if ( time() > $dismissal['last_dismissal'] + $this->next_show && $dismissal['number'] < $this->max_dismissals ) {
+				$score = 1;
+			}
+
+			return $score ? $score : false;
+		}
+
+		return $score;
+	}
+
+	/**
+	 * Score hosted with partner
+	 *
+	 * @param string $path path.
+	 * @param array  $query query.
+	 * @param int    $score score.
+	 *
+	 * @return bool|int
+	 */
+	private function score_hosted_with_partner( $path, $query, $score ) {
+		if ( is_null( $this->hosted_with_partner ) ) {
+			return $score;
+		}
+
+		// Using 'bluehost' for development
+		// TODO: replace with proper detection method.
+
+		if ( 'bluehost' === $this->hosted_with_partner ) {
+			return 1;
+		}
+
+		return false;
+	}
+
+	/**
+	 * Score user locale
+	 *
+	 * @param string $path path.
+	 * @param array  $query query.
+	 * @param int    $score score.
+	 *
+	 * @return bool|int
+	 */
+	private function score_user_locale( $path, $query, $score ) {
+		if ( empty( $this->user_locales ) ) {
+			return $score;
+		}
+
+		$user_locale = strtolower( get_user_locale() );
+
+		return in_array( $user_locale, $this->user_locales, true );
+	}
+
+	/**
+	 * Score user roles
+	 *
+	 * @param string $path path.
+	 * @param array  $query query.
+	 * @param int    $score score.
+	 *
+	 * @return bool|int
+	 */
+	private function score_user_roles( $path, $query, $score ) {
```

Comment:
I don't think we need this if we're only showing pre-connection JITMs to admins. Likewise with other scoring methods.